### PR TITLE
Add JENKINS_USE_GET_KUBE_SCRIPT option to support using get-kube.sh

### PR DESCRIPTION
**Pull request by ixdy**
Additionally include the get-kube.sh script in kubekins-e2e.

By default, this should have no effect on Jenkins. Testing in https://github.com/kubernetes/kubernetes/pull/35083 anyway. (Previous runs there set `JENKINS_USE_GET_KUBE_SCRIPT=y` to force that flow on PR Jenkins, and everything still worked.)

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/885)

### CODE REVIEW

**Comment by fejta** [comment]
What does get-kube.sh do?
The bash is hard to read and hard to unit test. How could we move most/all of this logic into golang or python?

**Comment by ixdy** [comment]
[`get-kube.sh`](https://github.com/kubernetes/kubernetes/blob/master/cluster/get-kube.sh) downloads the latest kubernetes release, extracts it, and starts a kubernetes cluster. It's equivalent to what's on https://get.k8s.io/. It's intentionally written with no dependencies so you can `curl ... | bash`.

(It's also been recently extended with logic to automatically download the appropriate client and server binaries if they aren't included in kubernetes.tar.gz, which is why I want to use it on Jenkins.)

**Comment by fejta** [comment]
Thanks!

What do you think about adding a [`--download` step](https://github.com/kubernetes/kubernetes/blob/master/hack/e2e.go#L51) to e2e.go or some such? Or even refactoring the acquiring the output binaries entirely out of e2e-runner.sh and into e2e.go now?

**Comment by spxtr** [comment]
Historical note: https://github.com/kubernetes/kubernetes/blob/fe76243b05b307ecda3138b01ff4d704f84d7d26/hack/e2e.go#L204-L250

**Comment by ixdy** [comment]
As a result of your comments, I'm now realizing that `stage_local_tars_and_set_release_vars` shouldn't be put here; we should fix the release scripts so that they create proper-looking local paths so no munging is necessary.

This appears to be easy to do, but will require changes in 3 repos (yay). I'll come back to this PR after that's done.

**Comment by spxtr** [comment]
Please reassign once it's ready for review.

**Comment by ixdy** [comment]
I've reworked this a bit after https://github.com/kubernetes/release/pull/166 and https://github.com/kubernetes/test-infra/pull/890. The logic to move tarballs around has been removed, but most of the rest remains.

tl;dr IMO this PR is about as simplistic as it can be right now. trying to move stuff into e2e.go only makes things more complicated and harder to understand. if/when we gut the rest of e2e-runner.sh, we can look into moving this stuff somewhere else. right now, it's a distraction.

longer explanation:
I've spent a while trying to figure out what can be moved into `hack/e2e.go`, but I'm not really sure it's worth it at this point.

The output of my current effort is https://github.com/ixdy/kubernetes/commit/fd470e7a65d04065639c1e7c4caaef32f4270618. This would basically replace 10 lines of shell in `set_release_vars_from_gcs` with 40 lines of Go, which doesn't strike me as being significantly better. We'd still need to call `hack/e2e.go` from this script, as well as ingest its output accordingly, probably resulting in several lines of different shell. `set_release_vars_from_gke_cluster_version` would still require special logic (to read the GKE default cluster version), and `set_release_vars_from_local_gcs_stage` would also still be a special case.

I thought about moving calling `get-kube.sh` into `e2e.go`, but that has its own issues; either we have to add all of these Jenkins-specific environment variables, or we have to add a bunch more flags (with shell code to translate the env vars into flags), and at the end of the day, we still need to set `CLUSTER_API_VERSION` somewhere. (This should probably be set automatically by e2e.go somewhere, but I don't want to shave that yak right now.)

There's also the fact that e2e.go expects to be run from a kubernetes/ directory, so using it to download and extract a kubernetes/ directory seems kinda weird, too.

File: jenkins/e2e-image/e2e-runner.sh [hunk 1]
```diff
@@ -84,6 +84,54 @@ function running_in_docker() {
...
+    if [[ -z "${KUBERNETES_RELEASE}" ]]; then
+      echo "FAIL! version not found in ${local_gcs_stage_path}"
+      return 1
+    fi
+}
+
+# Use a published version like "ci/latest" (default), "release/latest",
+# "release/latest-1", or "release/stable".
+# TODO(ixdy): maybe this should be in get-kube.sh?
+function set_release_vars_from_gcs() {
```
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: it may not be clear, but this function and the next one (for GKE) are just slightly modified versions of existing functions.

**Comment by ixdy** [comment]
FYI I tested this as-is in https://github.com/kubernetes/kubernetes/pull/35083 and everything worked as expected.

I'm testing it again with `JENKINS_USE_GET_KUBE_SCRIPT="y"` to see what happens.

**Comment by fejta** [comment]
How is someone supposed to test the behavior of pulling down tarballs outside of jenkins?

File: jenkins/e2e-image/Dockerfile [hunk 2]
```diff
@@ -45,10 +45,11 @@ ENV FAIL_ON_GCP_RESOURCE_LEAK=true \
 
 ADD ["e2e-runner.sh", \
     "kops-e2e-runner.sh", \
+    "https://raw.githubusercontent.com/kubernetes/kubernetes/master/cluster/get-kube.sh", \
```
Comment (fejta) [inline jenkins/e2e-image/Dockerfile]: Why is get-kube.sh in the kubernetes repo instead of test-infra?

File: jenkins/e2e-image/e2e-runner.sh [hunk 3]
```diff
@@ -84,6 +84,54 @@ function running_in_docker() {
     grep -q docker /proc/self/cgroup
 }
 
+### FUNCTIONS TO SUPPORT get-kube.sh ###
+
+# Moves local tarballs into a local path suitable for get-kube.sh, and sets
+# KUBERNETES_RELEASE and KUBERNETES_RELEASE_URL accordingly.
+function set_release_vars_from_local_gcs_stage() {
```
Comment (fejta) [inline jenkins/e2e-image/e2e-runner.sh]: We started added unit tests for non-trivial test-infra functionality. Which of these methods are interesting to unit test? What is the best way to add this unit test coverage?

**Comment by ixdy** [comment]
Humans should just run `cluster/get-kube.sh` (or `curl https://get.k8s.io` first) to pull release tarballs.

File: jenkins/e2e-image/Dockerfile (hunk 2 above)
Comment (ixdy) [inline jenkins/e2e-image/Dockerfile]: it's not specific to test-infra.

File: jenkins/e2e-image/e2e-runner.sh (hunk 3 above)
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: the logic is so trivial that I'm not sure any is.

Comment (fejta) [inline jenkins/e2e-image/e2e-runner.sh]: I do not consider a 400 line bash file trivial, and I would like to see the complexity of our bash files approach zero over time.

This PR seems to be moving in the other direction: nearly 100 lines of new bash code, lots of branches, state is maintained in global environment variables.

File: jenkins/e2e-image/e2e-runner.sh [hunk 4]
```diff
@@ -84,6 +84,54 @@ function running_in_docker() {
...
+    # Use latest build of the server version's branch for test files.
+    set_release_vars_from_gcs "ci/latest-${server_version:0:3}"
+}
+
+function call_get_kube() {
+    [[ "${JENKINS_USE_GET_KUBE_SCRIPT:-}" =~ ^[yY]$ ]] || exit 2
+    export KUBERNETES_RELEASE
+    export KUBERNETES_RELEASE_URL
+    KUBERNETES_SKIP_CONFIRM=y KUBERNETES_SKIP_CREATE_CLUSTER=y KUBERNETES_DOWNLOAD_TESTS=y \
+      "$(dirname "${0}")/get-kube.sh"
```
Comment (fejta) [inline jenkins/e2e-image/e2e-runner.sh]: Can we pass arguments into this program instead of relying on environment variables?

File: jenkins/e2e-image/e2e-runner.sh (hunk 3 above)
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: I should probably move most of `set_release_vars_from_gcs` to `get-kube.sh`, and possibly also write some unit tests for `get-kube.sh`. that is orthogonal to the intent of this PR, which is blocking my ability to pull out the client/server binaries from `kubernetes.tar.gz`.

File: jenkins/e2e-image/e2e-runner.sh [hunk 5]
```diff
@@ -273,25 +327,47 @@ EXIT_CODE=0
 # We get the Kubernetes tarballs unless we are going to use old ones
 if [[ "${JENKINS_USE_EXISTING_BINARIES:-}" =~ ^[yY]$ ]]; then
```
Comment (fejta) [inline jenkins/e2e-image/e2e-runner.sh]: If we keep this in bash I would like to see unit tests that validate various sets of environment variables copy files from X and into Y.

File: jenkins/e2e-image/e2e-runner.sh (hunk 3 above)
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: After I verify that using `get-kube.sh` works, L133 - L194 and L353 - L370 go away, as do all of the conditionals using `JENKINS_USE_GET_KUBE_SCRIPT`.

I'm duplicating the existing logic, guarded by this environment variable, since there's no better way to test Jenkins changes incrementally.

There will be less logic in this script when we're all done.

File: jenkins/e2e-image/e2e-runner.sh (hunk 4 above)
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: Not without breaking existing published functionality. The script is intended to be called like `curl https://get.k8s.io | bash`. Changing that is outside the scope of this PR.

File: jenkins/e2e-image/e2e-runner.sh (hunk 3 above)
Comment (fejta) [inline jenkins/e2e-image/e2e-runner.sh]: I get that, and my goal is to simultaneously a) unblock your pulling binaries from the tarball and b) either ensure that the complexity of this file decreases in the process or else the additional complexity is tested.

However if my opinion is an outlier on the team and the rest of our rotation is eager to support this I will be happy to defer to the consensus here.

**Comment by ixdy** [comment]
I know we want to kill `e2e-runner.sh`. I'm happy to support that.

I'm trying to get e2e testing setup around `cluster/get-kube.sh`. I need to do this so that I can make progress on improving the kubernetes release tarballs. Rewriting `e2e-runner.sh` in a different language is a distraction from that goal.

File: jenkins/e2e-image/e2e-runner.sh (hunk 5 above)
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: This isn't new.

File: jenkins/e2e-image/e2e-runner.sh (hunk 3 above)
Comment (ixdy) [inline jenkins/e2e-image/e2e-runner.sh]: The net effect of my change:
* we no longer pull tarballs from GCS in `e2e-runner.sh`
* we no longer extract tarballs in `e2e-runner.sh`

**Comment by ixdy** [comment]
`JENKINS_USE_GET_KUBE_SCRIPT="y"` apparently still works on PR Jenkins - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/35083/pull-kubernetes-e2e-gce/1230/?log

<details>
```
I1021 17:57:06.215] Cleaning up binaries.
W1021 17:57:06.219] + KUBERNETES_RELEASE=v1.5.0-alpha.1.757+842621e5034770
W1021 17:57:06.219] + [[ -z v1.5.0-alpha.1.757+842621e5034770 ]]
W1021 17:57:06.219] + call_get_kube
W1021 17:57:06.220] + [[ y =~ ^[yY]$ ]]
W1021 17:57:06.220] + export KUBERNETES_RELEASE
W1021 17:57:06.220] + export KUBERNETES_RELEASE_URL
W1021 17:57:06.220] ++ dirname /workspace/e2e-runner.sh
W1021 17:57:06.221] + KUBERNETES_SKIP_CONFIRM=y
W1021 17:57:06.221] + KUBERNETES_SKIP_CREATE_CLUSTER=y
W1021 17:57:06.221] + KUBERNETES_DOWNLOAD_TESTS=y
W1021 17:57:06.221] + /workspace/get-kube.sh
I1021 17:57:06.225] Downloading kubernetes release v1.5.0-alpha.1.757+842621e5034770
I1021 17:57:06.226]   from file:///workspace/_output/gcs-stage/v1.5.0-alpha.1.757+842621e5034770/kubernetes.tar.gz
I1021 17:57:06.226]   to /workspace/kubernetes.tar.gz
W1021 17:57:06.234] Warning: Illegal date format for -z, --timecond (and not a file name). 
W1021 17:57:06.235] Warning: Disabling time condition. See curl_getdate(3) for valid date syntax.
W1021 17:57:06.235]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
W1021 17:57:06.235]                                  Dload  Upload   Total   Spent    Left  Speed
W1021 17:57:06.670] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
 82  277M   82  229M    0     0   717M      0 --:--:-- --:--:-- --:--:--  717M
100  277M  100  277M    0     0   636M      0 --:--:-- --:--:-- --:--:--  415M
I1021 17:57:06.672] Unpacking kubernetes release v1.5.0-alpha.1.757+842621e5034770
I1021 17:57:09.031] Kubernetes release: v1.5.0-alpha.1.757+842621e5034770
I1021 17:57:09.031] Server: linux/amd64
I1021 17:57:09.031] Client: linux/amd64
I1021 17:57:09.031] 
I1021 17:57:09.032] Will download and extract kubernetes-test.tar.gz from file:///workspace/_output/gcs-stage/v1.5.0-alpha.1.757+842621e5034770
W1021 17:57:09.040]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
W1021 17:57:09.040]                                  Dload  Upload   Total   Spent    Left  Speed
W1021 17:57:09.311] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  164M  100  164M    0     0   608M      0 --:--:-- --:--:-- --:--:--  608M
I1021 17:57:09.311] 
I1021 17:57:09.655] md5sum(kubernetes-test.tar.gz)=88ffe42b4fb4352daa0eba88e03a551f
```
</details>

File: jenkins/e2e-image/Dockerfile (hunk 2 above)
Comment (spxtr) [inline jenkins/e2e-image/Dockerfile]: I think it's what you get when you do a `curl https://get.k8s.io/ | bash`. Not test-infra specific. I think it's kinda gross to include it in the `e2e-image`, but I don't know a better way to do this.

File: jenkins/e2e-image/e2e-runner.sh [hunk 6]
```diff
@@ -273,25 +327,47 @@ EXIT_CODE=0
...
-    # test what's running in GKE by default rather than some CI build.
-    clean_binaries
-    fetch_server_version_tars
-elif [[ "${JENKINS_USE_GCI_VERSION:-}" =~ ^[yY]$ ]]; then
-    clean_binaries
-    fetch_gci_version_tars
 else
-    # use JENKINS_PUBLISHED_VERSION, default to 'ci/latest', since that's
-    # usually what we're testing.
     clean_binaries
```
Comment (fejta) [inline jenkins/e2e-image/e2e-runner.sh]: I get that this eliminates a couple duplicate lines of clean_binaries, but I do not think the additional branch depth justifies this. Especially since it makes it harder to understand what you are adding.

**Comment by ixdy** [comment]
I haven't been super clear about how this fits into the overall plan.

Basically, I'd like to get architecture-specific binaries out of `kubernetes.tar.gz` before we enter 1.5 code slush. I've been working on a way to do this transparently without breaking anyone. Most downstream uses of the client and server binaries already use the -server or -client tarballs directly, but Jenkins doesn't, and https://get.k8s.io/ (which is just `cluster/get-kube.sh`) didn't until recently.

(Side note: https://get.k8s.io is designed to follow the pattern of a number of OSS tools, in which you can download/install by just running `curl -sS https://get.k8s.io/ | bash -`. It's targeted for users who want a more custom installation than is provided by one of the managed services (i.e. GKE), but not for someone who's comfortable just building their own release. As such, we can't expect them to have Python or Go installed, and so we much rely on bash alone.)

Now https://get.k8s.io looks for a `cluster/get-kube-binaries.sh` script in the `kubernetes.tar.gz` release tarball and runs it if it exists. This script is designed to automatically download release and test binaries as necessary.

I've decided to use this script in Jenkins since it removes one of the duplicated code paths for downloading release artifacts.
It also actually gives us some testing coverage on https://get.k8s.io (currently there's none), and will let me verify that the bootstrapping scripts I've written will actually work.

We can certainly move some of the logic around setting variables to pass into the script into Go. I really want to get this first effort done before the code slush, however, so Go-ifying existing scripts is not a huge priority right now.

I included `get-kube.sh` in the docker image because GitHub has been notoriously unreliable, and downloading it each time would likely lead to flakes.

The legacy code can be removed as soon as I verify that the `get-kube.sh` stuff all works correctly in PR and CI testing. I'm pretty sure it won't work on 1.2 or 1.3 right now, but I've already sent out cherry-pick PRs to fix this.
